File: April-Challenges-2021/Day-4/Design-Circular-Queue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.items = k * [None]
        self.maxSize = k
        self.start = -1
        self.top = -1

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False
        else:
            if self.top+1==self.maxSize:
                self.top = 0
            else:
                self.top+=1
                if self.start == -1:
                    self.start = 0
            self.items[self.top] = value 
            return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        else:
            firstElement = self.items[self.start]
            start = self.start
            if self.start == self.top:
                self.start = -1
                self.top = -1
            elif self.start+1 == self.maxSize:
                self.start = 0
            else:
                self.start += 1
            self.items[start] = None            
            # return firstElement
            return True
                

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        else:
            return self.items[self.start] 

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        else:
            return self.items[self.top]

    def isEmpty(self) -> bool:
        return self.top == -1  

    def isFull(self) -> bool:
        if self.top+1==self.start:
            return True
        elif self.start==0 and self.top+1==self.maxSize:
            return True
        else:
            return False


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()

File: Golang/Design-Circular-Queue.go
type Node struct {
    Val int
    Next *Node
}

type LinkedList struct {
    Head *Node
    Tail *Node
    Size int
}

func (this *LinkedList) Append(val int) {
    
    node := &Node{val, nil}
    
    if this.Head == nil {
        this.Head = node
        this.Tail = node
    } else {
        this.Tail.Next = node
        this.Tail = this.Tail.Next
    }
    
    this.Size++
}

func (this *LinkedList) Prepend(val int) {
    
    node := &Node{val, nil}
    
    if this.Head == nil {
        this.Head = node
        this.Tail = node
    } else {
        node.Next = this.Head
        this.Head = node
    }
    
    this.Size++
}

func (this *LinkedList) PopLeft() *Node {
    if this.Size>0 {
        temp := this.Head
        this.Head = this.Head.Next
        this.Size--
        return temp
    }
    return &Node{0,nil}
}

type MyCircularQueue struct {
    Queue *LinkedList
    Size int
    MaxSize int
}


func Constructor(k int) MyCircularQueue {
    
    queue := MyCircularQueue{&LinkedList{nil,nil,0}, 0, k}
    return queue
}


func (this *MyCircularQueue) EnQueue(value int) bool {
    if this.Size+1 <= this.MaxSize {
        this.Queue.Append(value)
        this.Size++
        return true
    }
    return false
}


func (this *MyCircularQueue) DeQueue() bool {
    if this.Size > 0 {
        this.Queue.PopLeft()
        this.Size--
        return true
    }
    return false
}


func (this *MyCircularQueue) Front() int {
    if this.Size > 0 {
        return this.Queue.Head.Val
    }
    return -1
}


func (this *MyCircularQueue) Rear() int {
    if this.Size > 0 {
        return this.Queue.Tail.Val
    }
    return -1
}


func (this *MyCircularQueue) IsEmpty() bool {
    return this.Size == 0
}


func (this *MyCircularQueue) IsFull() bool {
    return this.Size == this.MaxSize
}

// [1,2,3]
/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * obj := Constructor(k);
 * param_1 := obj.EnQueue(value);
 * param_2 := obj.DeQueue();
 * param_3 := obj.Front();
 * param_4 := obj.Rear();
 * param_5 := obj.IsEmpty();
 * param_6 := obj.IsFull();
 */
